contracts/concentrated/PrivatePool.sol

```diff
@@ -0,0 +1,308 @@
+        require(_token0Decimals <= 18, "token0 decimals > 18");
+        token0Decimals = _token0Decimals;
+
+        uint256 _token1Decimals = uint256(IERC20Metadata(_token1).decimals());
+        require(_token1Decimals <= 18, "token1 decimals > 18");
+        token1Decimals = _token1Decimals;
+    }
+
+    /// @notice Updates the quote price LP is willing to offer tokens at
+    /// @param _P The new fixed price LP is willing to buy and sell at
+    function quote(uint256 _P) external onlyOwner {
+        require(_P >= PRICE_MIN && _P <= PRICE_MAX, "price out of range");
+        require(_P != P, "same price");
+
+        // set new P price param
+        P = _P;
+
+        emit Quote(_P);
+    }
+
+    /// @notice Updates the fee applied on swaps
+    /// @dev Effectively acts as bid/ask spread for LP
+    /// @param _fee The new swap fee
+    function setSwapFee(uint256 _fee) external onlyOwner {
+        require(_fee <= FEE_MAX, "fee > max");
+        fee = _fee;
+        emit NewSwapFee(_fee);
+    }
+
+    /// @notice Adds liquidity to pool
+    /// @param amounts The token amounts to add in token decimals
+    /// @param minToMint The minimum amount of liquidity to be minted
+    /// @param deadline The deadline before which liquidity must be added
+    function addLiquidity(
+        uint256[] calldata amounts,
+        uint256 minToMint,
+        uint256 deadline
+    ) external onlyOwner deadlineCheck(deadline) returns (uint256 minted_) {
+        require(amounts.length == 2, "invalid amounts");
+
+        // get current token balances in pool
+        uint256 xWad = _amountWad(IERC20(token0).balanceOf(address(this)), true);
+        uint256 yWad = _amountWad(IERC20(token1).balanceOf(address(this)), false);
+
+        // get D balance before add liquidity
+        uint256 _d = _D(xWad, yWad);
+
+        // convert amounts to wad for calcs
+        uint256 amount0Wad = _amountWad(amounts[0], true);
+        uint256 amount1Wad = _amountWad(amounts[1], false);
+
+        // balances after transfer
+        xWad += amount0Wad;
+        yWad += amount1Wad;
+
+        // calc diff with new D value
+        minted_ = _D(xWad, yWad) - _d;
+        require(minted_ >= minToMint, "minted < min");
+        _d += minted_;
+
+        // transfer amounts in decimals in
+        IERC20(token0).safeTransferFrom(msg.sender, address(this), amounts[0]);
+        IERC20(token1).safeTransferFrom(msg.sender, address(this), amounts[1]);
+
+        uint256[] memory fees = new uint256[](2);
+        fees[0] = 0;
+        fees[1] = 0;
+
+        emit AddLiquidity(msg.sender, amounts, fees, _d, _d);
+    }
+
+    /// @notice Removes liquidity from pool
+    /// @param amount The amount of liquidity to remove
+    /// @param minAmounts The minimum amounts of token to receive in token decimals
+    /// @param deadline The deadline before which liquidity must be removed
+    function removeLiquidity(
```

Comment:
I think it makes sense to add the option to withdraw a single token as well. One way to do this is to implement `removeLiquidityOneToken` allowing the LP to remove arbitrary amount of any of the tokens.

This would help LP to rebalance the pool, and doing a single token withdrawal instead of
1. Remove full liquidity using `amount = D` in `removeLiquidity`
2. Re-add liquidity in different proportion

Reply:
Sounds good! Given the fixed price curve, think this should be relatively straightforward. Just remove the specified token, and `D()` should take care of the curve rebalance.

Reply:
Updated `removeLiquidity` to take in similar params to add liquidity, so users can specify token amounts `uint256[] amounts` as array to remove.

Reply:
This is in fact the big brain solution to this, nice!